handlers/utils/compute_utils.go

```diff
@@ -0,0 +1,1186 @@
+	mapstructure.Decode(createConfig, &config)
+	setupConfig(instance.Data["fields"].(map[string]interface{}), &config)
+	setupEnvs(createConfig, &config)
+	labels, ok := GetFieldsIfExist(instance.Data, "fields", "labels")
+	if ok {
+		setupLabels(labels.(map[string]interface{}), &config)
+	}
+
+	container := GetContainer(client, instance, false)
+	containerID := ""
+	if container != nil {
+		containerID = container.ID
+	}
+	created := false
+	if len(containerID) == 0 {
+		newID, createErr := createContainer(client, &config, &hostConfig, imageTag, instance, name, progress)
+		if createErr != nil {
+			logrus.Error(fmt.Sprintf("fail to create container error :%s", createErr.Error()))
+		} else {
+			containerID = newID
+			created = true
+		}
+	}
+	logrus.Info(fmt.Sprintf("Starting docker container [%s] docker id [%s] %v", name, containerID, startConfig))
+
+	startErr := client.ContainerStart(context.Background(), containerID, types.ContainerStartOptions{})
+
+	if startErr != nil {
+		if created {
+			if err1 := removeContainer(client, containerID); err1 != nil {
+				logrus.Error(err1)
+			}
+		}
+		logrus.Error(startErr)
+		return startErr
+	}
+
+	RecordState(client, instance, containerID)
+	return nil
+}
+
+func GetInstanceAndHost(event *events.Event) (*model.Instance, *model.Host) {
+
+	data := event.Data
+	var ihm model.InstanceHostMap
+	mapstructure.Decode(data["instanceHostMap"], &ihm)
+
+	var instance model.Instance
+	mapstructure.Decode(ihm.Instance, &instance)
+	var host model.Host
+	mapstructure.Decode(ihm.Host, &host)
+
+	clusterConnection, ok := GetFieldsIfExist(data, "field", "clusterConnection")
+	if ok {
+		host.Data["clusterConnection"] = InterfaceToString(clusterConnection)
+		if strings.HasPrefix(InterfaceToString(clusterConnection), "http") {
+			caCrt, ok1 := GetFieldsIfExist(event.Data, "field", "caCrt")
+			clientCrt, ok2 := GetFieldsIfExist(event.Data, "field", "clientCrt")
+			clientKey, ok3 := GetFieldsIfExist(event.Data, "field", "clientKey")
+			// what if we miss certs/key? do we have to panic or ignore it?
+			if ok1 && ok2 && ok3 {
+				host.Data["caCrt"] = InterfaceToString(caCrt)
+				host.Data["clientCrt"] = InterfaceToString(clientCrt)
+				host.Data["clientKey"] = InterfaceToString(clientKey)
+			} else {
+				logrus.Infof("Missing certs/key [%v]for clusterConnection for connection ",
+					clusterConnection)
+			}
+		}
+	}
+	return &instance, &host
+}
+
+func IsInstanceActive(instance *model.Instance, host *model.Host) bool {
+	if IsNoOp(instance.Data) {
+		return true
+	}
+
+	client := docker.GetClient(DefaultVersion)
+	container := GetContainer(client, instance, false)
+	return isRunning(client, container)
+}
+
+func IsNoOp(data map[string]interface{}) bool {
+	b, ok := GetFieldsIfExist(data, "processData", "containerNoOpEvent")
+	if ok {
+		return InterfaceToBool(b)
+	}
+	return false
+}
+
+func GetContainer(client *client.Client, instance *model.Instance, byAgent bool) *types.Container {
```

Comment:
python agent doesn't return an error because if there is an error it will return an nil pointer so that go agent know that it does't exist and then will create a new container.